llvm/test/Analysis/ScalarEvolution/ptrtoaddr-i32-index-width.ll

```diff
@@ -0,0 +1,70 @@
+; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
+; RUN: opt -passes='print<scalar-evolution>' -disable-output %s 2>&1 | FileCheck %s
+
+target datalayout="p:64:64:64:32"
+
+define void @ptrtoaddr(ptr %in, ptr %out0) {
+; CHECK-LABEL: 'ptrtoaddr'
+; CHECK-NEXT:  Classifying expressions for: @ptrtoaddr
+; CHECK-NEXT:    %p0 = ptrtoaddr ptr %in to i32
+; CHECK-NEXT:    --> %p0 U: full-set S: full-set
```

Comment:
For ptrtoaddr we *do* want to create SCEV expressions for these. Handling the case where the address size doesn't match the pointer size is important.

I think the only case we might want to reject for ptrtoaddr is the one with hasUnstableRepresentation.

I feel like for ptrtoaddr the whole distinction of "lossless ptrtoaddr" doesn't really make sense (given that the result width is fixed). I think we should have only getPtrToAddr(). We should share only the core implementation of getLosslessPtrToIntExpr, without the data layout checks.

Reply:
With the clarification to icmp semantics from https://github.com/llvm/llvm-project/pull/163936, it is also fine to use ptrtoaddr with mismatched pointer/address size in all the places that use getLosslessPtrToIntExpr() on icmps (like computeExitLimit and helpers).